File: V1.11/dbg/main.cpp
#include <thread>
#include <chrono>
#include <opencv2/opencv.hpp>
#include <sercom.h>

using namespace std::chrono_literals;
using namespace sercom;

#pragma pack(push, 1)

struct State
{
	uint16_t addr;
	uint8_t data;
	uint8_t flags;

	void print() const
	{
		const char* hex = "0123456789ABCDEF";

		for (int shift = 12; shift >= 0; shift -= 4)
			std::cout << hex[(addr >> shift) & 15];
		std::cout << "  ";

		for (int shift = 4; shift >= 0; shift -= 4)
			std::cout << hex[(data >> shift) & 15];
		std::cout << "  ";

		if ((flags & 0x10) != 0) std::cout << " M1";
		if ((flags & 0x08) != 0) std::cout << " WR";
		if ((flags & 0x04) != 0) std::cout << " RD";
		if ((flags & 0x02) != 0) std::cout << " MREQ";
		if ((flags & 0x01) != 0) std::cout << " IORQ";
		std::cout << std::endl;
	}
};

class StateMonitor
{
	std::shared_ptr<Serial>		m_Serial;
public:
	StateMonitor()
		: m_Serial(Serial::create("COM14",115200))
	{}

	State sample()
	{
		constexpr uint8_t MAGIC[] = { 0x12,0x43,0x78,0xFA };
		const uint8_t c=65;
		m_Serial->send(&c, 1);
		std::vector<uint8_t> buffer, cur;
		buffer.reserve(256);
		size_t pos = 0;
		size_t wait_count = 0;
		size_t fail_count = 0;
		while (true)
		{
			size_t n = m_Serial->data_available();
			if (n == 0)
			{
				std::this_thread::sleep_for(10ms);
				if (++wait_count >= 10)
				{
					wait_count = 0;
					if (++fail_count >= 10) throw std::runtime_error("Failed to sample state");
					m_Serial->send(&c, 1);
				}
			}
			else
			{
				cur.resize(n);
				m_Serial->receive(cur);
				buffer.insert(buffer.end(), cur.begin(), cur.end());
				while (buffer.size() >= 8)
				{
					bool failed = false;
					for (size_t i = 0; i < 4; ++i)
					{
						if (buffer[i] != MAGIC[i])
						{
							buffer.erase(buffer.begin(), buffer.begin() + i + 1);
							failed = true;
							break;
						}
					}
					if (!failed)
					{
						State res;
						res.addr = (uint16_t(buffer[5]) << 8) | buffer[4];
						res.data = buffer[6];
						res.flags = buffer[7];
						return res;
					}
				}
			}
		}
	}
};


int main(int argc, char* argv[])
{
	cv::namedWindow("w");
	try
	{
		auto dbg_serial = Serial::create("COM16", 115200);
		StateMonitor sm;
		int k = 0;
		while ((k=cv::waitKey(10))!=27)
		{
			size_t n = dbg_serial->data_available();
			if (n > 0)
			{
				std::string s(n, ' ');
				dbg_serial->receive((uint8_t*)&s[0], n);
				std::cout << s;
			}
			if (k == 32)
			{
				std::cout << "------------------------------\n";
				State s = sm.sample();
				s.print();
				const uint8_t c = 65;
				dbg_serial->send(&c, 1);
				std::this_thread::sleep_for(10ms);
				s = sm.sample();
				s.print();
			}
		}
	}
	catch (const std::exception& e)
	{
		std::cerr << e.what() << std::endl;
	}
	return 0;
}

File: V1.11/dbg/main.py
import sys

import argh
import numpy as np
import serial
import threading
import cv2

done = False


class StateReader:
    def __init__(self):
        self.ser = serial.Serial('COM14', 115200, timeout=1)
        self.terminating = False
        self.thread = threading.Thread(target=self.read_thread)
        self.thread.start()
        self.addr = 0
        self.data = 0
        self.flags = 0

    def shutdown(self):
        self.terminating = True
        self.thread.join()

    def print_state(self):
        flags = ''
        if (self.flags & 0x10) != 0:
            flags += " M1"
        if (self.flags & 0x08) != 0:
            flags += " WR"
        if (self.flags & 0x04) != 0:
            flags += " RD"
        if (self.flags & 0x02) != 0:
            flags += " MREQ"
        if (self.flags & 0x01) != 0:
            flags += " IORQ"
        s = f'{self.addr:04x} {self.data:02x} {flags}'
        print(s)

    @staticmethod
    def magic(buffer: bytearray):
        if len(buffer) < 4:
            return False
        return buffer[0] == 0x12 and buffer[1] == 0x43 and buffer[2] == 0x78 and buffer[3] == 0xFA

    def decode_state(self, buffer: bytearray):
        n = len(buffer)
        while n >= 8:
            if not self.magic(buffer):
                del buffer[0]
            else:
                self.addr = (buffer[4] << 8) | buffer[5]
                self.data = buffer[6]
                self.flags = buffer[7]
                del buffer[0:8]

    def read_thread(self):
        buffer = bytearray()
        while not self.terminating:
            data = self.ser.read(8)
            buffer.extend(data)
            self.decode_state(buffer)


def read_feedback(s: serial.Serial):
    while not done:
        data = s.readall()
        if len(data) > 0:
            sys.stdout.write(data.decode('ascii'))


def kbd():
    ctrl_ser = serial.Serial('COM16', 115200, timeout=0.1)
    feedback_thread = threading.Thread(target=read_feedback(ctrl_ser))
    feedback_thread.start()
    cv2.namedWindow('w')
    img = np.zeros((256, 256), dtype=np.uint8)
    cv2.imshow('w', img)
    while True:
        k = cv2.waitKey(100)
        if k == ord('q'):
            break
        if k == ord('m'):
            ctrl_ser.write(bytes([65] * 500))
            ctrl_ser.flush()
        if k == 32:
            ctrl_ser.write(b'C')
            ctrl_ser.flush()
    global done
    done = True
    feedback_thread.join()


def main():
    kbd_thread = threading.Thread(target=kbd)
    kbd_thread.start()
    sr = StateReader()
    while not done:
        sr.print_state()
    sr.shutdown()
    kbd_thread.join()


if __name__ == '__main__':
    argh.dispatch_command(main)
